packages/core/src/render3/instructions/listener.ts

```diff
@@ -192,12 +202,29 @@ function listenerInternal(
         const subscription = output.subscribe(listenerFn);
         const idx = lCleanup.length;
         lCleanup.push(listenerFn, subscription);
-        tCleanup && tCleanup.push(eventName, tNode.index, idx, -(idx + 1));
+        if (tCleanup) {
+          storeListenerTCleanup(tNode, tCleanup, eventName, tNode.index, idx, -(idx + 1));
+        }
       }
     }
   }
 }
 
+/**
+ * A utility function that stores information about listener cleanup functions in TView.cleanup data
+ * structures. Also updates TNode to mark start / end indexes of cleanup entries in TView.cleanup
+ * for a given TNode.
+ */
+function storeListenerTCleanup(
+    tNode: TNode, tCleanup: any[], eventName: string, idxOrTargetGetter: any, listenerFnIdx: number,
+    useCaptureOrIndx: boolean | number): void {
+  if (tNode.cleanupStart === -1) {
+    tNode.cleanupStart = tCleanup.length;
+  }
+  tCleanup.push(eventName, idxOrTargetGetter, listenerFnIdx, useCaptureOrIndx);
+  tNode.cleanupEnd = tNode.cleanupEnd + 4;
```

Comment:
Wouldn't `tNode.cleanupEnd` be `-1` initially? So the first time this runs, `tNode.cleanupStart` will be the starting`tCleanup.length` and `tNode.cleanUpEnd` will always be `3`.

```suggestion
  tNode.cleanupEnd = tNode.cleanupEnd < 0 ? tNode.cleanupStart + 4 : tNode.cleanupEnd + 4;
```

Seems like a test should have caught this?

Reply:
Good catch, thank you! Slightly changing your suggestion to compare with `-1` to be consistent with the checks above.